docs/custom_promise_types/groups.py

```diff
@@ -0,0 +1,335 @@
+
+
+class GroupsPromiseTypeModule(PromiseModule):
+    def __init__(self):
+        super().__init__("git_promise_module", "0.0.2")
+        self._name_regex = r"[a-z_][a-z0-9_-]*[$]?\Z"
+        self._name_maxlen = 32
+
+
+    def validate_promise(self, promiser, attributes):
+        # check promiser type
+        if type(promiser) is not str:
+            raise ValidationError("invalid type for promiser: expected string")
+
+        # check promiser value
+        match = re.match(self._name_regex, promiser)
+        if match == None:
+            self.log_warning(f"promiser groupname '{promiser}' should match regular expression '[a-z_][a-z0-9_-]*[$]?'")
+
+        # check promiser length not too long
+        if len(promiser) > self._name_maxlen:
+            raise ValidationError(f"promiser '{promiser}' is too long: ({len(promiser)} > {self._name_maxlen})")
+
+        # if attribute policy is present
+        if "policy" in attributes:
+            policy = attributes["policy"]
+
+            # check attribute policy type
+            if type(policy) is not str:
+                raise ValidationError("invalid type for attribute policy: expected string")
+
+            # check attribute policy value
+            if policy != "present" and policy != "absent":
+                raise ValidationError(f"invalid value '{policy}' for attribute policy: must be 'present' or 'absent'")
+
+            # check attributes gid and members are not used with policy absent
+            if policy == "absent":
+                if "gid" in attributes:
+                    self._log_warning(f"cannot assign gid to absent group '{promiser}'")
+                if "members" in attributes:
+                    self._log_warning(f"cannot assign members to absent group '{promiser}'")
+        
+        # if attribute gid is present
+        if "gid" in attributes:
+            gid = attributes["gid"]
+
+            # check attribute gid type
+            if type(gid) is not str:
+                raise ValidationError("invalid type for attribute gid: expected string")
+
+            # check attribute gid value
+            if type(gid) == str:
+                try: 
+                    int(gid)
+                except ValueError:
+                    raise ValidationError(f"invalid value '{gid}' for attribute gid: expected integer literal")
+
+        # if attribute members is present
+        if "members" in attributes:
+            # some black magic which removes back slashes parsing before parsing json
+            members = json.loads(json.loads('"' + attributes["members"] + '"'))
+            attributes["members"] = members
+            
+            # check attribute only not used with attributes include or exclude
+            if "only" in members and ("include" in members or "exclude" in members):
+                raise ValidationError("attribute only may not be used with attributes include or exclude")
```

Comment:
I'd add single quotes around the attribute names to make the message easier to read.